isisd/isis_spf_delay.c

```diff
@@ -0,0 +1,480 @@
+  area = reinit->area;
+  level = reinit->level;
+  family = reinit->family;
+  assert(area);
+
+  if (family == AF_INET) {
+    family = 0;
+  } else if (family == AF_INET6) {
+    family = 1; 
+  }
+  
+  area->spf_delay_ietf[level-1]->family[family]->state = ISIS_SPF_DELAY_STATE_LONG_WAIT;
+  timerclear(&area->spf_delay_ietf[level-1]->family[family]->ttl_time);	
+  area->spf_delay_ietf[level-1]->family[family]->t_timetolearn = NULL;	
+  if (isis->debugs & DEBUG_SPF_IETF) {
+    zlog_debug ("ISIS-Spf (%s) L%d IETF SPF delay TIMETOLEARN expired moved to state %d (%s) family %d\n",
+                area->area_tag, 
+		level, 
+		area->spf_delay_ietf[level-1]->family[family]->state,
+                isis_spf_delay_states[area->spf_delay_ietf[level-1]->family[family]->state],
+		family);
+  }
+
+  return 1;
+}
+
+
+int
+isis_spf_delay_event_holddown_expires(struct thread *thread) {
+
+  int level = 0; 
+  struct isis_spf_delay_reinit	*reinit;	
+  struct isis_area	*area;
+  int family = 0;
+
+  reinit = THREAD_ARG(thread);
+  assert(reinit);
+  area = reinit->area;
+  level = reinit->level;	
+  family = reinit->family;
+	
+  assert(area);
+  if (family == AF_INET) {
+    family = 0;
+  } else if (family == AF_INET6) {
+    family = 1;
+  }
+
+  if (isis->debugs & DEBUG_SPF_IETF)
+    zlog_debug ("ISIS-Spf (%s) L%d F%d IETF SPF delay : disabling timetolearn (thread %p)",
+                area->area_tag, 
+		level, 
+		family, 
+		area->spf_delay_ietf[level-1]->family[family]->t_timetolearn);
+
+  THREAD_TIMER_OFF(area->spf_delay_ietf[level-1]->family[family]->t_timetolearn);
+  area->spf_delay_ietf[level-1]->family[family]->t_holddown = NULL;
+  area->spf_delay_ietf[level-1]->family[family]->t_timetolearn = NULL;
+  timerclear(&area->spf_delay_ietf[level-1]->family[family]->first_event_time);
+  timerclear(&area->spf_delay_ietf[level-1]->family[family]->resettime);
+  timerclear(&area->spf_delay_ietf[level-1]->family[family]->ttl_time);
+  area->spf_delay_ietf[level-1]->family[family]->state = ISIS_SPF_DELAY_STATE_QUIET;
+
+  if (isis->debugs & DEBUG_SPF_IETF) {
+    zlog_debug ("ISIS-Spf (%s) L%d IETF SPF delay HOLDDOWN expired moved to state %d (%s) family %d\n",
+		area->area_tag, 
+		level, 
+		area->spf_delay_ietf[level-1]->family[family]->state,
+		isis_spf_delay_states[area->spf_delay_ietf[level-1]->family[family]->state],
+		family);
+  }
+	
+  return 1;
+}
+
+int
+isis_spf_delay_ietf_schedule(struct isis_area *area, int level, int family) {
+  struct timeval time_now;
+  long current_wait_time;
+  struct isis_spf_delay_reinit  *reinit;
+  unsigned char old_spf_state;
+
+  reinit = (struct isis_spf_delay_reinit *)malloc(sizeof(struct isis_spf_delay_reinit));
+  old_spf_state = area->spf_delay_ietf[level-1]->family[family]->state;
+ 
+  if (family != 0) 
+   return ISIS_OK; 
+  
+  reinit->area = area;
+  reinit->level = level;
+  reinit->family = AF_INET;
+
+  assert (area->is_type & level);
+
+  gettimeofday(&time_now, NULL);
```

Comment:
Using `gettimeofday` is incorrect in almost all cases, particularly in all cases that use relative timing (which this is). It will break horribly if the system starts up without correct clock and then at some point finishes NTP sync, making the clock jump. Not to mention, it also breaks for leap seconds.

`monotime()` is the correct clock to use here.